pubsub/client.go

```diff
@@ -161,116 +132,131 @@ func (c *Client) Listener(kind ListenerType, event string) *Listener {
 	return listener
 }
 
-// Tears down the client - closes the connection and stops listening for
-// connections.
-func (c *Client) TearDown() {
-	c.setState(ClosedState)
-
-	if c.pubsub != nil {
-		c.pubsub.Close()
-	}
-}
-
 // Gets the current client state.
-func (c *Client) GetState() State {
+func (c *Client) GetState() uint8 {
 	c.stateLock.Lock()
 	defer c.stateLock.Unlock()
 
-	return c.State
+	return c.state.State()
 }
 
 // Sets the client state in a thread-safe manner.
-func (c *Client) setState(s State) {
+func (c *Client) setState(s uint8) error {
 	c.stateLock.Lock()
-	c.State = s
+	err := c.state.Goto(s)
 	c.stateLock.Unlock()
+
+	if err != nil {
+		return err
+	}
+
+	switch s {
+	case ConnectedState:
+		c.emit(ConnectedEvent, nil)
+	case DisconnectedState:
+		c.emit(DisconnectedEvent, nil)
+	case ClosingState:
+		c.emit(ClosingEvent, nil)
+	case ClosedState:
+		c.emit(ClosedEvent, nil)
+	}
+
+	return nil
 }
 
-// Blocks until a certain event type is emitted on the client.
-func (c *Client) WaitFor(event EventType) Event {
-	for {
-		if ev := <-c.Events; ev.Type == event {
-			return ev
-		}
+// Tries to reconnect to pubsub, looping until we're able to do so
+// successfully. This must be called to activate the client.
+func (c *Client) Connect() {
+	for c.GetState() == DisconnectedState {
+		go c.resubscribe()
+		c.doConnection()
+		time.Sleep(c.policy.Next())
```

Comment:
I'm not sure that's really appropriate here. It's a single delay, not an interval.